### DISCUSSION TITLE
references support?

### DISCUSSION THREAD

**Discussion by steffenschumacher** [netbox-community/netbox#6849] [Ideas]
I have a use case where I map old to new switch ports when migrating access switches.

I need this, since I'm usually not migrating ports 1:1 - eg. 4 switches being replaced by 2 due to reduced usage, and ports are configured explicitly with differing vlans, and I also want to remove patches from unused ports.

To support this, in my current homegrown system I can reference a planned replacement switch from active switches, and reference a mapped/replacement port on the planned switch from each of the ports on the active switches.

I've added constraints etc. to ensure that only ports on the planned switch can be referenced from a to-be-decommissioned switch, and that its ports can only be referenced once.
I also use this when there are several standalone switches in a single cabinet, which will be replaced with a new stack of switches.
These references allows me to do several things:

- plan which ports are to be replaced by which new ports
- generate/transfer interface configurations for the replacement switch, including description, mode, vlans etc
- generate patch labels to assist in the replacement, indicating:
  - the old hostname+port
  - the new hostname+port 
  - and also info about the vlan
  - I use http://mr-label.com/articles/1107 and have some code to generate matching PDFs
- cache port state (link, macs learned in which vlans) and compare pre- with post-migration

I'm not sure this would be a feature worth the effort to implement into netbox, as I guess it is quite bespoke, and statically configured vlans are slowly going away with IBNS/similar - but possibly it could be supported with custom fields, where a new type of custom field is introduced - a reference to an entry in table? 
It would probably be complex to support also, but beyond mapping devices to devices and interfaces to interfaces, there could be loads of other stuff you'd want to map. It would probably have limited support for constraints, but simple stuff like controlling if the referenced value must be unique.

I could do something with plain custom fields, but that would leave me without database integrity.

Should I do a feature request for:

- Custom field references
- The specific feature I just described

Or just accept the status quo and live without database integrity?
(as a side note - we are replacing 1000s of switches, which is why we have put this much effort into these guardrails/training wheels for our operations)

**Comment by julianze** [comment]
I think a simple plugin could address this in the best way. 

View - select device:
A list of devices which has to be replaced.
Adding new devices for replacement planning.

View - Interface list:
 Thinking about a side-by-side table on the left side a list of interface 
of the switch/router which has to be replaced. 
 This should be easy retrieved by the dcim.Interface model.
 For every interface on the left side you can choose on the right side a 
switch and an interface from this selected switch.


 When selecting the new interface a submit button can save this planned 
reference in a new table as status "planned" and will put a "reserved - replacement sw-X Gig0/1" in the description of the new interface.
Additional marking this interface as connected will help, that this port will not be "stolen" by others.

Second action "stage" could dump all setting from old to new interface, so an automation can now prepare the new port with vlans etc.

Last action "Patch" is the physical part of a remote hand to switchover the cable from old to new. Behind netbox will deactivate the old port and optional swap the cable from old to new. But this makes only senses when the same cable is used (same rack).

**Comment by julianze** [comment]
In additional this could be interesting for us. We dont have to replace thousand switches, but replacements are a common task in our operations team.

**Reply by steffenschumacher** [reply]
So we would have some budget for this, but don’t care about the UI part - I don’t have any idea about the cost of having this developed, nor if there are developers experienced with these plugins to be had?
But if this plug-in is done, I guess it should ideally be supported so that it will work in future releases also, so possibly two parties supporting this is not enough?

**Reply by sdktr** [reply]
Got questioned about the exact same use case this week. This was in a broader sense then just the workflow of moving assets (interface settings/IP addresses), but also at which point in time these actions should be executed. If done in advance the config for the destination device can be prepped based on the SoT. But in the meanwhile the still active device is not in sync with the SoT.

It would be ideal if netbox would allow different-point-in-time views of the data. Eg an IP-adres is 'Active' on interface X, but the same IP is allowed to attach to interface Y as 'Planned'.
